### DISCUSSION TITLE
Multiple statement execution

### DISCUSSION THREAD

**Discussion by khushboovashi** [psycopg/psycopg#417] [General]
In psycopg3, If I execute multiple statements in a single query, I will get the result of the first statement. 
For example, cursor.execute('SELECT 1; SELECT 2;'), will give the result 1. 
In psycopg2, the result will be 2, the last statement will be considered here.

So, is this intensional ? What is the benefit of this ?

**Comment by dvarrazzo** [comment]
Yes, it is intentional. In Psycopg 3 you can access all the results in a multiple statement, using [`Cursor.nextset()`](https://www.psycopg.org/psycopg3/docs/api/cursors.html#psycopg.Cursor.nextset).

Actually I see that mentioning this is missing from the [2 to 3 docs page](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#multiple-statements-in-the-same-query): will add.

**Reply by khushboovashi** [reply]
Thanks for the reply. I have tested the cursor.nextset() and It is working fine with multiple select queries, but 
If I execute below statements, cursor.nextset() is showing True, but when I try to fetch the records it gives me an error, psycopg.ProgrammingError: the last operation didn't produce a result 


cursor.execute("""DROP TABLE IF EXISTS test_for_notices;

DO $$
BEGIN
    RAISE NOTICE 'Hello, worldP!';
END $$;

SELECT 'CHECKING Multiple statements';""")

**Reply by dvarrazzo** [reply]
You have executed 3 statements, and only the third returns tuples, so you need two `nextset()` to get there.
```

In [20]: cur.statusmessage
Out[20]: 'DROP TABLE'

In [21]: cur.nextset()
Out[21]: True

In [22]: cur.statusmessage
Out[22]: 'DO'

In [23]: cur.nextset()
Out[23]: True

In [24]: cur.statusmessage
Out[24]: 'SELECT 1'

In [25]: cur.fetchone()
Out[25]: ('CHECKING Multiple statements',)
```

If you want to have a 100% sure way to know if `fetch*()` will return a result, you should check the status of the result, to be TUPLES_OK, not COMMAND_OK or something else:

```

In [30]: cur.pgresult.status == psycopg.pq.ExecStatus.TUPLES_OK
Out[30]: False

In [31]: cur.nextset()
Out[31]: True

In [32]: cur.pgresult.status == psycopg.pq.ExecStatus.TUPLES_OK
Out[32]: False

In [33]: cur.nextset()
Out[33]: True

In [34]: cur.pgresult.status == psycopg.pq.ExecStatus.TUPLES_OK
Out[34]: True

In [35]: cur.fetchone()
Out[35]: ('CHECKING Multiple statements',)
```

**Reply by khushboovashi** [reply]
Thanks for the help. This feature will benefit pgAdmin a lot.

**Reply by dvarrazzo** [reply]
I have added [a note in the docs](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#multiple-results-returned-from-multiple-statements) about the difference.
